File: src/Backend/ForgottenRealms.Simulation/GameLogic/GameEntities/Village.cs
using ForgottenRealms.Simulation.GameLogic.Interfaces;
using ForgottenRealms.Simulation.GameLogic.Models;
using ForgottenRealms.Simulation.SimulationModels;

namespace ForgottenRealms.Simulation.GameLogic.GameEntities;

public class Village : IGameEntity
{
    public Coordinates Coordinates { get; set; } = new() { X = 0, Y = 0 };
    public Resources Resources { get; set; } = new(){ Wood = 750, Clay = 750, Iron = 750 };
    public Resources ResourcesPerHour { get; set; } = new() { Wood = 20, Clay = 20, Iron = 20 };
    public Dictionary<BuildingEnum, int> BuildingLevels { get; set; } = new()
    {
        { BuildingEnum.Headquarters, 1 },
        { BuildingEnum.Woodcutter, 0 },
        { BuildingEnum.ClayPit, 0 },
        { BuildingEnum.IronMine, 0 },
    };
    public Task Update(ServerTick tick, CancellationToken cancellationToken)
    {
        var interval = tick.TickDuration.TotalHours;
        Resources.Wood += ResourcesPerHour.Wood * interval;
        Resources.Clay += ResourcesPerHour.Clay * interval;
        Resources.Iron += ResourcesPerHour.Iron * interval;
        return Task.CompletedTask;
    }

    public override string ToString() => $"Village [{Coordinates.X}|{Coordinates.Y}] - Wood: {Resources.Wood}, Clay: {Resources.Clay}, Iron: {Resources.Iron}";
}

File: src/Backend/forgotten-realms-server/src/game/models/village.rs
use crate::simulation::models::simulation_entity::SimulationEntity;

pub struct Village {
    coord_x: i32,
    coord_y: i32,
    wood: i32,
    clay: i32,
    iron: i32,
    wood_per_hour: i32,
    clay_per_hour: i32,
    iron_per_hour: i32,
}

impl Village {
    pub fn new(x: i32, y: i32) -> Self {
        Village {
            coord_x: x,
            coord_y: y,
            wood: 0,
            clay: 0,
            iron: 0,
            wood_per_hour: 100,
            clay_per_hour: 100,
            iron_per_hour: 100,
        }
    }
}

impl SimulationEntity for Village {
    fn update(&mut self) {
        self.wood += self.wood_per_hour / 3600;
        self.clay += self.clay_per_hour / 3600;
        self.iron += self.iron_per_hour / 3600;
    }
}